src/posix/platform/rcp_caps_diag.cpp

```diff
@@ -616,6 +620,88 @@ bool RcpCapsDiag::IsSpinelCapabilitySupported(const uint8_t *aCapsData, spinel_s
     return ret;
 }
 
+void RcpCapsDiag::HandleDiagOutput(const char *aFormat, va_list aArguments, void *aContext)
+{
+    static_cast<RcpCapsDiag *>(aContext)->HandleDiagOutput(aFormat, aArguments);
+}
+
+void RcpCapsDiag::HandleDiagOutput(const char *aFormat, va_list aArguments)
+{
+    int rval;
+
+    VerifyOrExit(mDiagOutput != nullptr && mDiagOutputLength != 0);
+    rval = vsnprintf(mDiagOutput, mDiagOutputLength, aFormat, aArguments);
+    VerifyOrExit(rval >= 0);
+
+    rval = (rval > mDiagOutputLength) ? mDiagOutputLength : rval;
+    mDiagOutput += rval;
+    mDiagOutputLength -= rval;
+
+exit:
+    return;
+}
+
+void RcpCapsDiag::ProcessSpinelSpeed(void)
+{
+    static constexpr uint32_t kUsPerSec           = 1000000;
+    static constexpr uint8_t  kBitsPerByte        = 8;
+    static constexpr uint8_t  kSpinelHeaderSize   = 4;
+    static constexpr uint8_t  kZeroTerminatorSize = 1;
+    static const char *const  kEchoCmd            = "echo ";
+    static constexpr uint8_t  kEchoPayloadLength  = 200;
+    static constexpr uint8_t  kNumTests           = 100;
+
+    otError                  error                                            = OT_ERROR_NONE;
+    char                     cmd[OPENTHREAD_CONFIG_DIAG_CMD_LINE_BUFFER_SIZE] = {0};
+    char                     output[OPENTHREAD_CONFIG_DIAG_OUTPUT_BUFFER_SIZE];
+    uint16_t                 i;
+    uint16_t                 echoPayloadLength;
+    uint64_t                 startTimestamp;
+    uint64_t                 endTimestamp;
+    uint64_t                 sumTime   = 0;
+    uint64_t                 sumLength = 0;
+    uint64_t                 speed;
+    otPlatDiagOutputCallback callback;
+    void                    *context;
+
+    SuccessOrExit(error = mRadioSpinel.GetDiagOutputCallback(&callback, &context));
+    mRadioSpinel.SetDiagOutputCallback(HandleDiagOutput, this);
+
+    strncpy(cmd, kEchoCmd, sizeof(cmd) - 1);
+    echoPayloadLength = sizeof(cmd) - strlen(cmd) - 1;
+    echoPayloadLength = (kEchoPayloadLength < echoPayloadLength) ? kEchoPayloadLength : echoPayloadLength;
+    memset(cmd + strlen(cmd), '1', echoPayloadLength);
+
+    for (i = 0; i < kNumTests; i++)
+    {
+        output[0]         = '\0';
+        mDiagOutput       = output;
+        mDiagOutputLength = sizeof(output);
+        startTimestamp    = otPlatTimeGet();
+
+        SuccessOrExit(error = mRadioSpinel.PlatDiagProcess(cmd));
+
+        endTimestamp = otPlatTimeGet();
+        sumTime += endTimestamp - startTimestamp;
+        sumLength += kSpinelHeaderSize + strlen(cmd) + kZeroTerminatorSize + kSpinelHeaderSize + strlen(output) +
+                     kZeroTerminatorSize;
+    }
+
+    mRadioSpinel.SetDiagOutputCallback(callback, context);
+
+exit:
+    if (error == OT_ERROR_NONE)
+    {
+        speed = (sumLength * kBitsPerByte * kUsPerSec) / sumTime;
+        snprintf(output, sizeof(output), "%lu bps", speed);
+        OutputFormat("SpinelSpeed", output);
```

Comment:
I expect this to give us a good estimate for UART, but does this calculation work for SPI as well? I am not sure.

Reply:
I don't expect this diag command could accurately estimate the speed of the UART or SPI. There are other factors that can greatly affect the Spinel speed, such as the delay in Thread HAL. I'd prefer to use this diag command to estimate the speed differences between different products or different spinel interface configurations.